src/shared/lib/MutationUtils.tsx

```diff
@@ -565,3 +571,66 @@ export function createAnnotatedProteinImpactTypeFilter(
     };
     return filter;
 }
+
+export function createNumericalFilter(getData: (d: Mutation) => number | null) {
+    const filter = (filter: NumericalFilter, mutation: Mutation) => {
+        const value = getData(mutation);
+        const lowerBound = filter.values[0];
+        const upperBound = filter.values[1];
+        return value !== null && value >= lowerBound && value <= upperBound;
+    };
+    return filter;
+}
+
+export function createCategoricalFilter(getData: (d: Mutation) => string) {
+    const filter = (filter: CategoricalFilter, mutation: Mutation) => {
+        const value = getData(mutation);
+        if (filter.values.length < 2 || value === '') {
+            return false;
+        }
+
+        const filterType = filter.values[0] as string;
+        const filterString = filter.values[1] as string;
+        const selections = filter.values[2] as Set<string>;
+        let match =
+            filterString === '' ||
+            matchCategoricalFilterSearch(
+                value.toUpperCase(),
+                filterType,
+                filterString.toUpperCase()
+            );
+        match = match && selections.has(value);
+        return match;
+    };
+    return filter;
+}
+
+export function matchCategoricalFilterSearch(
+    value: string,
+    filterType: string,
+    filterString: string
+) {
+    switch (filterType) {
+        case 'contains':
```

Comment:
I would define a separate function for each `DataFilter.type` instead of a lengthy switch case.

Reply:
Are we sure about this? This would require introducing 9 filter appliers for each categorical `MutationTableColumnType` (of which there are well over 10). Additionally, adding/removing/renaming any categorical filter types would require adding/removing/renaming that many more filter appliers for each column. I would also need to make some adjustments to the mutation mapper code in order to track the categorical filter type for each column, which seems confusing in my opinion, since I don't do that for numerical filters.

Why not just keep the categorical filter type as a pseudo-value called `filterCondition` as part of the `CategoricalFilterValue` object? I see it as just another data filter configuration option, like `filterString` or `lowerBound` and `upperBound` for the numerical filter. The alternative just seems too bloated and not extensible.

Reply:
I see. This one is a bit tricky. I guess we can think of it as `filterCondition` and pair it with the filter value object. I would still prefer a simple filter applier function doing a single thing, but I think I can live with this for now 🙂